Add Liveness, Readiness, and Startup Probes concept

pegasas opened the pull request:
> https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/ is more of a tutorial than a task.
> 
> If someone has the capacity to, we could split it into:
> 
> * a new tutorial page that walks people through learning about container probes
> * a focused guide for people who have already learned, and want to set up one of the four types of probe
>   
>   * with signposting to the tutorial, in case someone wants to learn the background first

pegasas commented:
trying to fix https://github.com/kubernetes/website/issues/42371

tengqm commented on content/en/docs/concepts/configuration/liveness-readiness-startup-probes.md:
```diff
@@ -0,0 +1,44 @@
+* TCP Socket
+* gRPC Protocol
+
+Read about [Configure Liveness, Readiness and Startup Probes](/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes).
```
Some feedbacks:

- You may need a English native speaker to help review the language;
- You may actually migrate some texts about basic concepts from the task page;
- You may want to emphasis on the use cases, the problems that are targeted by
  each type of probes, the design considerations behind the implementation that
  users may need to be aware of.

tengqm replied:
Also, please follow the structure for a concept page.

pegasas replied:
Sure.
I will make another version.

tengqm replied:
FYI. I have just proposed a fix to the existing probe task (https://github.com/kubernetes/website/pull/42619).

sftim commented:
This looks like a work in progress
/retitle [WIP] Add Liveness, Readiness, and Startup Probes concept

sftim commented:
@pegasas can we help you to move this work forward?

pegasas commented:
I have updated the PR according to your comments, 
feel free to review this.

kbhawkey commented on content/en/docs/concepts/configuration/liveness-readiness-startup-probes.md:
```diff
@@ -0,0 +1,43 @@
+
+Readiness probes determine when a container is ready to start accepting traffic. This is useful when waiting for an application to perform time-consuming initial tasks, such as establishing network connections, loading files, and warming caches. 
+
+If the readiness probe returns a failed state, the Kubernetes will remove the pod from all matching service endpoints.
```
If the readiness probe returns a failed state, Kubernetes removes the pod from all matching service endpoints.

pegasas replied:
Done.

kbhawkey commented on content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:
```diff
@@ -8,6 +8,8 @@ weight: 140
 
 This page shows how to configure liveness, readiness and startup probes for containers.
 
+You may wants to have an overview of [Liveness, Readiness and Startup Probes](/docs/concepts/configuration/liveness-readiness-startup-probes)
```
Can you reword the sentence to something like:
For an overview about probes, see [Liveness, Readiness, and Startup Probes](/docs/concepts/configuration/liveness-readiness-startup-probes).

kbhawkey commented:
Page preview:
https://deploy-preview-42376--kubernetes-io-main-staging.netlify.app/docs/concepts/configuration/liveness-readiness-startup-probes/

kbhawkey commented on content/en/docs/concepts/configuration/liveness-readiness-startup-probes.md:
```diff
@@ -0,0 +1,43 @@
+
+<!-- overview -->
+
+There are currently 3 types of probes in Kubernetes:
```
Kubernetes has various types of probes:

kbhawkey commented on content/en/docs/concepts/configuration/liveness-readiness-startup-probes.md:
```diff
@@ -0,0 +1,43 @@
+
+## Startup Probe
+
+A startup probe verifies whether the application within a container is started. This can be used to adopt liveness checks on slow starting containers, avoiding them getting killed by the kubelet before they are up and running.
```
nit: Does it make sense to capitalize the probe names?

divya-mohan0209 commented:
Gentle reminder @pegasas. Please advise if you have the capacity to work on moving the PR forward by addressing comments.

pegasas commented:
> Gentle reminder @pegasas. Please advise if you have the capacity to work on moving the PR forward by addressing comments.

Sure.
I think I've handle most of comments.
I will recheck to make it pass gated today.

sftim reviewed (changes requested):
I spotted some obvious technical inaccuracy. Please check the article matches how Kubernetes works; if you're not sure how to do that, then instead you should mention that as context for reviewers.

Where you spot that the proposed changes are misleading or clearly incorrect, please revise the text accordingly.

sftim commented:
(apologies for the time it took to explain why we can't merge this change)

We'd expect to see progress towards a merge. A reviewer is welcome to close this PR if we don't see updates before April 2024.

divya-mohan0209 reviewed (changes requested):
@pegasas : Thank you for all your work on this PR. We would like to understand if you have bandwidth to continue your efforts by addressing the suggestions provided by @sftim .

pegasas commented:
> @pegasas : Thank you for all your work on this PR. We would like to understand if you have bandwidth to continue your efforts by addressing the suggestions provided by @sftim .

It seems I missed the comments commits.
updated.

divya-mohan0209 commented on content/en/docs/concepts/configuration/liveness-readiness-startup-probes.md:
```diff
@@ -0,0 +1,45 @@
+
+## Liveness probe
+
+Liveness probes determine when to restart a container. For example, liveness probes could catch a deadlock, where an application is running, but unable to make progress.
```
Grammatical nit.
```suggestion
Liveness probes determine when to restart a container. For example, liveness probes could catch a deadlock, when an application is running, but unable to make progress.
```

divya-mohan0209 reviewed (commented):
Small grammatical nit, but otherwise, this LGTM!

divya-mohan0209 reviewed (commented):
Thanks for this, @pegasas !
/lgtm

nate-double-u reviewed (commented):
Thanks for this, @pegasas. You've addressed all the change requests, so I'll approve this even though the last one has yet to be officially removed. Thanks for your patience!

/approve
